jerry-core/ecma/base/ecma-lcache.c

```diff
@@ -79,241 +79,175 @@ ecma_lcache_init (void)
-                                                                    ecma_lcache_hash_table[hash_key][i].prop_name_cp);
+                                                                    entry_p->prop_name_cp);
 
       JERRY_ASSERT (prop_name_p->hash == entry_prop_name_p->hash);
 
       if (ECMA_STRING_GET_CONTAINER (prop_name_p) == ECMA_STRING_GET_CONTAINER (entry_prop_name_p)
           && prop_name_p->u.common_field == entry_prop_name_p->u.common_field)
       {
-        ecma_property_t *prop_p = ecma_lcache_hash_table[hash_key][i].prop_p;
-        JERRY_ASSERT (prop_p == NULL || ecma_is_property_lcached (prop_p));
+        ecma_property_t *prop_p = entry_p->prop_p;
+        JERRY_ASSERT (prop_p != NULL && ecma_is_property_lcached (prop_p));
 
         *prop_p_p = prop_p;
-
         return true;
       }
       else
       {
-        /* may be equal but it is long to compare it here */
+        /* They can be equal, but generic string comparison is too costly. */
       }
     }
+    entry_p++;
   }
-#else /* CONFIG_ECMA_LCACHE_DISABLE */
-  (void) object_p;
-  (void) prop_name_p;
-  (void) prop_p_p;
 #endif /* !CONFIG_ECMA_LCACHE_DISABLE */
 
   return false;
 } /* ecma_lcache_lookup */
 
 /**
  * Invalidate LCache entries associated with given object and property name / property
- *
- * Note:
- *      Either property name argument or property argument should be NULL,
- *      and another should be non-NULL.
- *      In case property name argument is NULL, property's name is taken
- *      from property's description.
  */
 void
 ecma_lcache_invalidate (ecma_object_t *object_p, /**< object */
-                        ecma_string_t *prop_name_p, /**< property's name (See also: Note) */
-                        ecma_property_t *prop_p) /**< property (See also: Note) */
+                        ecma_string_t *prop_name_p, /**< property's name */
+                        ecma_property_t *prop_p) /**< property */
 {
   JERRY_ASSERT (object_p != NULL);
   JERRY_ASSERT (prop_name_p != NULL);
+  JERRY_ASSERT (prop_p != NULL && ecma_is_property_lcached (prop_p));
+  JERRY_ASSERT (ECMA_PROPERTY_GET_TYPE (prop_p) == ECMA_PROPERTY_TYPE_NAMEDDATA
+                || ECMA_PROPERTY_GET_TYPE (prop_p) == ECMA_PROPERTY_TYPE_NAMEDACCESSOR);
 
 #ifndef CONFIG_ECMA_LCACHE_DISABLE
-  if (prop_p != NULL)
-  {
-    JERRY_ASSERT (ECMA_PROPERTY_GET_TYPE (prop_p) == ECMA_PROPERTY_TYPE_NAMEDDATA
-                  || ECMA_PROPERTY_GET_TYPE (prop_p) == ECMA_PROPERTY_TYPE_NAMEDACCESSOR);
+  jmem_cpointer_t object_cp;
+  ECMA_SET_NON_NULL_POINTER (object_cp, object_p);
 
-    bool is_cached = ecma_is_property_lcached (prop_p);
+  ecma_lcache_hash_entry_t *entry_p = ecma_lcache_hash_table[ecma_lcache_row_idx (object_cp, prop_name_p)];
 
-    if (!is_cached)
+  for (uint32_t entry_index = 0; entry_index < ECMA_LCACHE_HASH_ROW_LENGTH; entry_index++)
+  {
+    if (entry_p->object_cp == object_cp && entry_p->prop_p == prop_p)
```

Comment:
I was thinking about it, and moving the NULL to prop_p is inefficient because the lookup() function depends on the object_cp. Here a (entry_p->prop_p == prop_p) check is not enough, since a property can be added, removed, and added to another entry position. This could work though:

    if (entry_p->object_cp != ECMA_NULL_POINTER && entry_p->prop_p == prop_p)

But I think the difference is minor compared to the other one. I will try it though.